### ISSUE TITLE
link Domain to LDAP exception

### ISSUE THREAD

**Issue by izenk** [apache/cloudstack#2843]
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
Management server
~~~

##### CLOUDSTACK VERSION
~~~
4.11.1
~~~

##### CONFIGURATION
Advanced networking

##### OS / ENVIRONMENT
CentOS Linux release 7.4.1708

##### SUMMARY
Functionality: **link Domain to LDAP**
LDAP settings are configured
CloudStack see ldap and I can add users through 'Add LDAP account'
LDAP users can login

##### STEPS TO REPRODUCE
Create Domain -> Click '**link Domain to LDAP**' -> fill form -> click 'Ok'

##### EXPECTED RESULTS
~~~
Domain is linked to LDAP
~~~

##### ACTUAL RESULTS
Error: can not link a domain unless a basedn is configured for it
~~~
2018-09-13 11:24:48,503 TRACE [o.a.c.a.BaseCmd] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) Caught exception while checking first class entities for display property, continuing on
java.lang.ClassCastException: java.lang.Class cannot be cast to java.lang.String
        at org.apache.cloudstack.api.BaseCmd.isDisplay(BaseCmd.java:375)
        at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:109)
        at com.cloud.api.ApiServer.queueCommand(ApiServer.java:732)
        at com.cloud.api.ApiServer.handleRequest(ApiServer.java:556)
        at com.cloud.api.ApiServlet.processRequestInContext(ApiServlet.java:310)
        at com.cloud.api.ApiServlet$1.run(ApiServlet.java:130)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:56)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:103)
        at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.runWithContext(DefaultManagedContext.java:53)
        at com.cloud.api.ApiServlet.processRequest(ApiServlet.java:127)
        at com.cloud.api.ApiServlet.doGet(ApiServlet.java:89)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:527)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:530)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
        at java.lang.Thread.run(Thread.java:748)
2018-09-13 11:24:48,503 TRACE [c.c.u.d.T.Transaction] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) conn: Using existing DB connection
2018-09-13 11:24:48,503 TRACE [c.c.u.d.T.Statement] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) Preparing: SELECT domain_details.id, domain_details.domain_id, domain_details.name, domain_details.value FROM domain_details WHERE domain_details.domain_id = ?  AND domain_details.name = ?  ORDER BY RAND() LIMIT 1
2018-09-13 11:24:48,503 TRACE [c.c.u.d.T.Statement] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) Closing: com.mysql.jdbc.JDBC4PreparedStatement@381e9ba9: SELECT account.id, account.account_name, account.type, account.role_id, account.domain_id, account.state, account.removed, account.cleanup_needed, account.network_domain, account.uuid, account.default_zone_id, account.default FROM account WHERE account.id = 1
2018-09-13 11:24:48,504 TRACE [c.c.u.n.NioConnection] (AgentManager-NioConnectionHandler-1:null) (logid:) Keys Processing: 0
2018-09-13 11:24:48,504 TRACE [c.c.u.n.NioConnection] (AgentManager-NioConnectionHandler-1:null) (logid:) Keys Done Processing.
2018-09-13 11:24:48,506 INFO  [c.c.a.ApiServer] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) can not link a domain unless a basedn is configured for it.
2018-09-13 11:24:48,508 TRACE [c.c.a.r.ApiResponseSerializer] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) ===Serializing Response===
2018-09-13 11:24:48,509 DEBUG [c.c.a.ApiServlet] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) ===END===  192.168.7.78 -- GET  command=linkDomainToLdap&response=json&domainid=0fb4bbcc-a30a-4f5c-ab0a-e118ed4bbbc0&type=GROUP&name=demosso&accounttype=0&_=1536830688458
2018-09-13 11:24:48,509 TRACE [o.a.c.c.CallContext] (qtp1595953398-12:ctx-be9c3a72 ctx-7980a1a6) (logid:338bd4c9) Unregistered: CCtxt[acct=15; user=17; id=7980a1a6-2a33-4818-9fbe-ef8e05b3f03e]
2018-09-13 11:24:48,509 TRACE [c.c.u.d.T.Statement] (qtp1595953398-12:ctx-be9c3a72) (logid:338bd4c9) Closing: com.mysql.jdbc.JDBC4PreparedStatement@538c8b62: SELECT domain_details.id, domain_details.domain_id, domain_details.name, domain_details.value FROM domain_details WHERE domain_details.domain_id = 7  AND domain_details.name = _binary'ldap.basedn'  ORDER BY RAND() LIMIT 1
2018-09-13 11:24:48,509 TRACE [c.c.u.d.T.Connection] (qtp1595953398-12:ctx-be9c3a72) (logid:338bd4c9) Closing DB connection: dbconn516534065
2018-09-13 11:24:48,509 TRACE [c.c.u.d.T.Transaction] (qtp1595953398-12:ctx-be9c3a72) (logid:338bd4c9) Transaction is done
2018-09-13 11:24:48,509 TRACE [o.a.c.c.LogContext] (qtp1595953398-12:ctx-be9c3a72) (logid:338bd4c9) Unregistered: LogCtxt[acct=1; user=1; id=338bd4c9-badf-4beb-a4fe-ed39d1d4dece]
2018-09-13 11:24:48,509 TRACE [o.a.c.c.CallContext] (qtp1595953398-12:ctx-be9c3a72) (logid:) Unregistered: CCtxt[acct=1; user=1; id=be9c3a72-1172-4e1a-9241-1228e7c9d49a]
2018-09-13 11:24:48,513 TRACE [c.c.u.d.T.Transaction] (Cluster-Heartbeat-1:ctx-4bb657db) (logid:16cdf692) txn: DB Changes committed. Time = 27
2018-09-13 11:24:48,513 TRACE [c.c.u.d.T.Statement] (Cluster-Heartbeat-1:ctx-4bb657db) (logid:16cdf692) Closing: com.mysql.jdbc.JDBC4PreparedStatement@7b3fdd3a: update mshost set last_update='2018-09-13 09:24:48', removed=null, alert_count=0 where id=2 and runid=1536830612204
2018-09-13 11:24:48,513 TRACE [c.c.u.d.T.Connection] (Cluster-Heartbeat-1:ctx-4bb657db) (logid:16cdf692) Closing DB connection: dbconn324315536
~~~

**Comment by DaanHoogland** [comment]
@izenk I have not reproduced it, but I might have missed your point; did you enter the basedn on global level or not at all?

**Comment by izenk** [comment]
@DaanHoogland 
basedn is set on global level

**Comment by izenk** [comment]
this is SQL request (from management TRACE log):
~~~
mysql> SELECT domain_details.id, domain_details.domain_id, domain_details.name, domain_details.value FROM domain_details WHERE domain_details.domain_id = 7  AND domain_details.name = _binary'ldap.basedn'  ORDER BY RAND() LIMIT 1;
+----+-----------+-------------+----------------------------------------------------------+
| id | domain_id | name        | value                                                    |
+----+-----------+-------------+----------------------------------------------------------+
| 44 |         7 | ldap.basedn | TMJ1lT6Lqa/35kG9EV00OQOd8H0Tl0UDFQIjgpZqpgCHxrwjmIpEQQ== |
+----+-----------+-------------+----------------------------------------------------------+
1 row in set (0.00 sec)
~~~

**Comment by DaanHoogland** [comment]
I linked the domain to ldap without any of the basedn set, using head of the 4.11 branch, can you check that?

**Comment by izenk** [comment]
@DaanHoogland 
sorry, can't test

By error message I founded src here (**can not link a domain unless a basedn is configured for it**):
~~~
https://github.com/apache/cloudstack/blob/1d05fead49f5c856257a741b07122f5633d2e359/plugins/user-authenticators/ldap/src/main/java/org/apache/cloudstack/ldap/LdapManagerImpl.java
~~~

~~~
    @Override
    public LinkDomainToLdapResponse linkDomainToLdap(LinkDomainToLdapCmd cmd) {
        Validate.isTrue(_ldapConfiguration.getBaseDn(cmd.getDomainId()) == null, "can not link a domain unless a basedn is configured for it.");
        Validate.notEmpty(cmd.getLdapDomain(), "ldapDomain cannot be empty, please supply a GROUP or OU name");
        return linkDomainToLdap(cmd.getDomainId(),cmd.getType(),cmd.getLdapDomain(),cmd.getAccountType());
}
~~~

I upgraded cloudstack installation from 4.11.0 through yum REPO (http://cloudstack.apt-get.eu/centos/$releasever/4.11/)

**Comment by DaanHoogland** [comment]
ok @izenk, i'll try your scenario on 4.11.1 and see if there is a difference.

**Comment by DaanHoogland** [comment]
I can confirm this behabviour is different in 4.11.1 than 4.11-head. It's solved along the way after releasing 4.11.1. please wait for this till 4.11.2 comes out (and help testing that release ;)
